File: ruby/6kyu/sort_inner_content.rb
# You have to sort the inner content of every word of a string in descending order.
# The inner content is the content of a word without first and the last char.

# Some examples:

# "sort the inner content in descending order" -> "srot the inner ctonnet in dsnnieedcg oredr"
# "wait for me" -> "wiat for me"
# "this kata is easy" -> "tihs ktaa is esay"
# The string will never be null and will never be empty.
# It will contain only lowercase-letters and whitespaces.

def sort_the_inner_content(words)
  result = []
  
  words.split(" ").each do |word|
    new_string = []
    if word.length > 3
      new_string << word[0]
      new_string << word[1..-2].split("").sort.reverse.join("")
      new_string << word[-1]
    else
      new_string << word
    end
    result << new_string
  end
  result.map { |array| array.join("") }.join(" ")
end

# p sort_the_inner_content("sort the inner content in descending order")
# "srot the inner ctonnet in dsnnieedcg oredr"

# p sort_the_inner_content("wait for me")
# "wiat for me"

# p sort_the_inner_content("this kata is easy")
# "tihs ktaa is esay"

File: javascript/6kyu/sortInnerContent.js
/*
You have to sort the inner content of every word of a string
in descending order.
The inner content is the content of a word without first and
the last char.

Some examples:

"sort the inner content in descending order" -> "srot the
inner ctonnet in dsnnieedcg oredr"
"wait for me" -> "wiat for me"
"this kata is easy" -> "tihs ktaa is esay"
*/

function sortDescending(string) {
  const descending = (a, b) => b.localeCompare(a);
  return string.split('').sort(descending).join('');
}

function setInnerContent(string) {
  return string.length >= 2 ? sortDescending(string) : string;
}

function sortTheInnerContent(string) {
  return string.split(' ').map(word => {
    const firstLetter = word[0];
    if (word.length === 1) return firstLetter;
    const lastLetter = word[word.length - 1];
    let finalInnerContent = setInnerContent(word.slice(1, word.length -1));

    return `${firstLetter}${finalInnerContent}${lastLetter}`;
  }).join(' ');
}

console.log(sortTheInnerContent("sort the inner content in descending order"));
// "srot the inner ctonnet in dsnnieedcg oredr"

console.log(sortTheInnerContent('xsrolhew nn ivtqnfecbq nusrqhhu wjc kt ozysroomkq zqfddr pzvrdh'));
// 'xsrolhew n ivtqnfecbq nusrqhhu wjc kt ozysroomkq zqfddr pzvrdh'

console.log(sortTheInnerContent('ezzxohw cmjfdcav ayuuphgfj ssig uxsr ytfcx vv azrem'));
// 'ezzxohw cmjfdcav ayuuphgfj ssig uxsr ytfcx v azrem'